packages/notifications/src/elements/toaster/ToastProvider.tsx

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright Zendesk, Inc.
+ *
+ * Use of this source code is governed under the Apache License, Version 2.0
+ * found at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+import React, { createContext, useReducer, Dispatch, useCallback } from 'react';
+import PropTypes from 'prop-types';
+
+import {
+  toasterReducer,
+  getInitialState,
+  IToasterState,
+  ToasterReducerAction,
+  ToastPlacement
+} from './reducer';
+import { ToastSlot } from './ToastSlot';
+
+export const ToastContext = createContext<
+  { state: IToasterState; dispatch: Dispatch<ToasterReducerAction> } | undefined
+>(undefined);
+
+export interface IToastProviderProps {
+  /**
+   * Limit the number of visible toasts
+   */
+  limit?: number;
+  /**
+   * Sets the `z-index` of the toast
+   */
+  zIndex?: number;
+}
+
+export const ToastProvider: React.FC<IToastProviderProps> = ({ limit, zIndex, children }) => {
+  const [state, dispatch] = useReducer(toasterReducer, getInitialState());
+
+  const contextValue = { state, dispatch };
+
+  const toastsByPlacement = useCallback(
+    (placement: ToastPlacement) => {
+      let matchingToasts = state.toasts.filter(toast => toast.options.placement === placement);
+
+      if (placement === 'bottom' || placement === 'bottom-start' || placement === 'bottom-end') {
+        matchingToasts = matchingToasts.reverse();
+      }
+
+      return (
+        <ToastSlot placement={placement} toasts={matchingToasts} zIndex={zIndex} limit={limit!} />
+      );
+    },
+    [limit, state.toasts, zIndex]
+  );
+
+  return (
+    <ToastContext.Provider value={contextValue}>
+      {toastsByPlacement('top-start')}
+      {toastsByPlacement('top')}
+      {toastsByPlacement('top-end')}
+      {children}
+      {toastsByPlacement('bottom-start')}
+      {toastsByPlacement('bottom')}
+      {toastsByPlacement('bottom-end')}
+    </ToastContext.Provider>
+  );
+};
+
+ToastProvider.displayName = 'ToastProvider';
+
+ToastProvider.defaultProps = {
+  limit: 4
```

Comment:
How can the limit be made unlimited? I could definitely be wrong, but it feels like `undefined` should be the default.

Reply:
@m-lai originally raised concerns about readability with more than 4 items which is what led us to the default. "unlimited" seems a unique use-case. If that was something that needed to occur the user could provide [Number.POSITIVE_INFINITY](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/POSITIVE_INFINITY).

Reply:
Got it – makes sense.